lib/std/enums.zig

```diff
@@ -0,0 +1,1281 @@
+    const E = extern enum { a, b, c, d = 0, e, f = 3 };
+    testing.expectEqualSlices(E, &.{.a, .b, .c, .f}, uniqueValues(E));
+
+    const F = enum { a, b, c };
+    testing.expectEqualSlices(F, &.{.a, .b, .c}, uniqueValues(F));
+}
+
+/// Returns the set of all unique field values in the given enum, in
+/// declaration order.  For repeated values in extern enums, only the
+/// first name for each value is included.
+pub fn uniqueFields(comptime E: type) []const EnumField {
+    comptime {
+        const info = @typeInfo(E).Enum;
+        const raw_fields = info.fields;
+        // Only extern enums can contain duplicates,
+        // so fast path other types.
+        if (info.layout != .Extern) {
+            return raw_fields;
+        }
+
+        var unique_fields: []const EnumField = &[_]EnumField{};
+    outer:
+        for (raw_fields) |candidate| {
+            for (unique_fields) |u| {
+                if (u.value == candidate.value)
+                    continue :outer;
+            }
+            unique_fields = unique_fields ++ &[_]EnumField{candidate};
+        }
+
+        return unique_fields;
+    }
+}
+
+/// Determines the length of a direct-mapped enum array, indexed by
+/// @intCast(usize, @enumToInt(enum_value)).  The enum must be exhaustive.
+/// If the enum contains any fields with values that cannot be represented
+/// by usize, a compile error is issued.  The max_unused_slots parameter limits
+/// the total number of items which have no matching enum key (holes in the enum
+/// numbering).  So for example, if an enum has values 1, 2, 5, and 6, max_unused_slots
+/// must be at least 3, to allow unused slots 0, 3, and 4.
+fn directEnumArrayLen(comptime E: type, comptime max_unused_slots: comptime_int) comptime_int {
+    const info = @typeInfo(E).Enum;
+    if (!info.is_exhaustive) {
+        @compileError("Cannot create direct array of non-exhaustive enum "++@typeName(E));
+    }
+
+    var max_value: comptime_int = -1;
+    const max_usize: comptime_int = ~@as(usize, 0);
+    const fields = uniqueFields(E);
+    for (fields) |f| {
+        if (f.value < 0) {
+            @compileError("Cannot create a direct enum array for "++@typeName(E)++", field ."++f.name++" has a negative value.");
+        }
+        if (f.value > max_value) {
+            if (f.value > max_usize) {
+                @compileError("Cannot create a direct enum array for "++@typeName(E)++", field ."++f.name++" is larger than the max value of usize.");
+            }
+            max_value = f.value;
+        }
+    }
+
+    const unused_slots = max_value + 1 - fields.len;
+    if (unused_slots > max_unused_slots) {
+        const unused_str = std.fmt.comptimePrint("{d}", .{unused_slots});
+        const allowed_str = std.fmt.comptimePrint("{d}", .{max_unused_slots});
+        @compileError("Cannot create a direct enum array for "++@typeName(E)++". It would have "++unused_str++" unused slots, but only "++allowed_str++" are allowed.");
```

Comment:
This should probably return an error rather than a `@compileError`, otherwise a user cannot provide a fallback as an enum grows.

Reply:
I don't understand the use case for catching this error.  This is a safety check to make sure that you are not accidentally making an array with a large number of empty slots.  If this triggers, the way to handle it is to inspect it and decide whether to increase the allowed number of unused slots or switch to a denser packing like EnumArray.  If you are writing generic code that calls this function, your code should accept this parameter alongside the enum.

Can you provide a more specific example where you would want to catch this?